partridge/cli.py

```diff
@@ -0,0 +1,136 @@
+from collections import defaultdict
+
+import click
+import numpy as np
+import pandas as pd
+from tqdm import tqdm
+
+from .config import default_config
+from .gtfs import feed as mkfeed
+from .hmm import (
+    CandidateCache,
+    ObservationCache,
+    MeasureCache,
+    NoPathError,
+    _shape_dist_traveled,
+)
+from .utilities import remove_node_attributes
+from .writers import write_feed_dangerously
+
+
+ALIGNMENT_COL = '__odometer_idx__'
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.command()
+@click.argument('infile')
+@click.argument('outfile')
+def odometer(infile, outfile):
+    '''
+    A tool for measuring the distance traveled by a vehicle.
+    '''
+    feed = mkfeed(infile, config=_read_config())
+    outfeed = mkfeed(infile, config=_write_config())
+
+    click.echo('Read trips.txt')
+    feed.trips
+
+    click.echo('Read stop_times.txt')
+    feed.stop_times
+
+    click.echo('Read stops.txt')
+    feed.stops
+
+    click.echo('Read shapes')
+    feed.shapes
+
+    click.echo('Prepare row alignments for writing')
+    feed.stop_times[ALIGNMENT_COL] = feed.stop_times.index
+    outfeed.stop_times[ALIGNMENT_COL] = outfeed.stop_times.index
+
+    click.echo('Sort and group shapes')
+    shapes = (
+        feed.shapes
+            .sort_values('shape_pt_sequence')
+            .groupby('shape_id'))
+
+    click.echo('Merge, sort, and group observations')
+    trips = (
+        feed.stop_times
+            .merge(feed.trips, sort=False)
+            .merge(feed.stops, sort=False)
+            .sort_values('stop_sequence')
+            .groupby('trip_id'))
+
+    candidates = {}
+    for shape_id, points in tqdm(shapes, desc='Prune and cache HMM states'):
+        shape_lons = points.shape_pt_lon.values
+        shape_lats = points.shape_pt_lat.values
+        if shape_lons.shape[0] < 2:
+            continue
+        if pd.isnull(shape_lons).any() or pd.isnull(shape_lats).any():
+            continue
+        candidates[shape_id] = CandidateCache(shape_lons, shape_lats)
+
+    observations = {}
+    row_alignments = defaultdict(list)
+    for trip_id, obs in tqdm(trips, desc='Prune and cache HMM observations'):
+        stop_lons = obs.stop_lon.values
+        stop_lats = obs.stop_lat.values
+        shape_id = obs.iloc[0].shape_id
+        stop_ids = tuple(obs.stop_id)
+        if obs.shape[0] < 2:
+            continue
+        if shape_id not in candidates:
+            continue
+        if pd.isnull(stop_lons).any() or pd.isnull(stop_lats).any():
+            continue
+        if stop_ids not in observations:
+            observations[stop_ids] = ObservationCache(stop_lons, stop_lats)
+        row_alignments[shape_id, stop_ids].append(np.array(obs[ALIGNMENT_COL]))
+
+    if not any(row_alignments):
+        click.echo('No opportunities to generate shape_dist_traveled.')
```

Comment:
Also, might be interesting to report back that, say, `8 out 12 candidates can have shape_dist calculated`.